meetings/working-groups/collection-literals/CL-2024-01-23.md

```diff
@@ -0,0 +1,143 @@
+
+### Extension methods
+
+We are still interesting in enabling `[a, b, c].ExtensionMethod()`. However, we think this is not something that should be limited to collection expressions.  Broadly speaking, the language is very limited wrt to the interaction of extension methods and target-typed constructs.  For example, you cannot do this `(x => true).ExtensionOnStringPredicate()`.  The language requires the item being invoked have a type *prior* to lookup of the extension method.  To make cases like this work, we'd need to allow them to not have a type, and then say that lookup should still find the extensions, which are then tested for applicability in the rewritten form (e.g. `ExtensionMethod([a, b, c])`).
+
+As we are investing heavily in extensions in C#13, we may want to roll the exploration of this area into that work.
+
+### Addressing existing types that should work with collection expressions, but don't.
+
+There are several types we've found that annoying don't work with collection expressions.  For example:
+
+- ArraySegment<T>
+- ReadOnlyMemory<T>.  Commonly used as the analog of ReadOnlySpan<T> when you're working with async methose.
+- InlineArray
+
+We want to continue collecting these types to see what would be a good approach to expanding out support in the future.  Specifically, if the set of types is very small, it might be acceptable to just hardcode support for them.  However, if the set grows large, we may need to identify patterns to allow them to participate.  For example, we might say that if a type has a `public static implicit operator ThisType(T[] array)` conversion operator, that it would then be constructible with a collection expression.
+
+It's worth noting that making any type now be collection expression constructible would always be a potential breaking change with overload resolution, unless we made these always have a lower priority than the core rules we shipped with.
+
+### Supporting non-generic interfaces
+
+There is an open question on what the (currently illegal) semantics should be for `IEnumerable x = [1, 2, 3];`
+
+There are a few paths we could conceivable take here.
+
+- Have no special behavior for this, and allow 'natural type' to take care of it.  If, for example, the natural type of `[1, 2, 3]` was `List<int>` then this would be equivalent to: `IEnumerable x = (List<int>)[1, 2, 3];`.  Note that this would then be making a mutable collection under the covers, strongly typed to 'int'.
+
+- Treat the non-generic interfaces as the generic versions with 'object'.  In this case, that would mean the above would be equivalent to: `IEnumerable x = (IEnumerable<object>)[1, 2, 3];`.  This would then mean the compiler would synthesize the read-only collection, but also that every element would be boxed as object in restful storage.
```

Comment:
```suggestion
- Treat the non-generic interfaces as the generic versions with 'object'.  In this case, that would mean the above would be equivalent to: `IEnumerable x = (IEnumerable<object>)[1, 2, 3];`.  This would then mean the compiler would synthesize the read-only collection, but also that every element would be boxed as object.